pkg/distributor/distributor.go

```diff
@@ -340,129 +376,188 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
+			// series we're trying to dedupe when HA tracking moves over to a different replica.
+			if removeReplica {
+				removeLabel(d.limits.HAReplicaLabel(userID), &ts.Labels)
+			}
 
-		if len(ts.Labels) == 0 {
-			continue
-		}
+			for _, labelName := range d.limits.DropLabels(userID) {
+				removeLabel(labelName, &ts.Labels)
+			}
 
-		// We rely on sorted labels in different places:
-		// 1) When computing token for labels, and sharding by all labels. Here different order of labels returns
-		// different tokens, which is bad.
-		// 2) In validation code, when checking for duplicate label names. As duplicate label names are rejected
-		// later in the validation phase, we ignore them here.
-		sortLabelsIfNeeded(ts.Labels)
+			if len(ts.Labels) == 0 {
+				continue
+			}
 
-		// Generate the sharding token based on the series labels without the HA replica
-		// label and dropped labels (if any)
-		key, err := d.tokenForLabels(userID, ts.Labels)
-		if err != nil {
-			return nil, err
-		}
+			// We rely on sorted labels in different places:
+			// 1) When computing token for labels, and sharding by all labels. Here different order of labels returns
+			// different tokens, which is bad.
+			// 2) In validation code, when checking for duplicate label names. As duplicate label names are rejected
+			// later in the validation phase, we ignore them here.
+			sortLabelsIfNeeded(ts.Labels)
+
+			// Generate the sharding token based on the series labels without the HA replica
+			// label and dropped labels (if any)
+			key, err := d.tokenForLabels(userID, ts.Labels)
+			if err != nil {
+				return nil, err
+			}
 
-		validatedSeries, err := d.validateSeries(ts, userID)
+			validatedSeries, err := d.validateSeries(ts, userID)
 
-		// Errors in validation are considered non-fatal, as one series in a request may contain
-		// invalid data but all the remaining series could be perfectly valid.
-		if err != nil {
-			lastPartialErr = err
-		}
+			// Errors in validation are considered non-fatal, as one series in a request may contain
+			// invalid data but all the remaining series could be perfectly valid.
+			if err != nil {
+				lastPartialErr = err
+			}
 
-		// validateSeries would have returned an emptyPreallocSeries if there were no valid samples.
-		if validatedSeries == emptyPreallocSeries {
-			continue
+			// validateSeries would have returned an emptyPreallocSeries if there were no valid samples.
+			if validatedSeries == emptyPreallocSeries {
+				continue
+			}
+
+			seriesKeys = append(seriesKeys, key)
+			validatedTimeseries = append(validatedTimeseries, validatedSeries)
+			validatedSamples += len(ts.Samples)
 		}
+	}
 
-		keys = append(keys, key)
-		validatedTimeseries = append(validatedTimeseries, validatedSeries)
-		validatedSamples += len(ts.Samples)
+	if len(req.Metadata) > 0 {
+		for _, m := range req.Metadata {
+			key := d.tokenForMetadata(userID, m.MetricName)
```

Comment:
Can you move it after the `if err != nil {` so that we don't generate it if the validation fails?